fix: fix spark/sql test failures in native_iceberg_compat

parthchandra opened the pull request:
## Which issue does this PR close?

Part of #1542 

## Rationale for this change

A bug in the logic of `NativeBatchReader` caused NPE and array index out of bounds errors in `native_iceberg_compat` mode. Summary is that the old version used `requestedSchema.getColumns` to get the columns to read. However, this returns only the leaf (primitive) columns and does not contain any group fields.  So if the query was trying to read a group field (i.e. reading an entire struct instead of just one of the fields of the struct), we would use incorrect column metadata and sometimes even an incorrect number of fields.

## What changes are included in this PR?
The PR changes the logic to use `requestedSchema.getFields` which returns both group and primitive type fields.
The PR also adds additional handling in the `schema_adapter` to allow fields in the `to_type` schema that may not exist in the `from_type` schema 


## How are these changes tested?

New unit tests (based on the tests that were failing in Spark).

parthchandra commented:
@mbutrovich @andygrove this is ready for review.

parthchandra commented on common/src/main/java/org/apache/comet/parquet/NativeBatchReader.java:
```diff
@@ -496,25 +526,25 @@ private int loadNextBatch() throws Throwable {
-      columnReaders[i] = reader;
+      if (!missingColumns[i]) {
+        if (columnReaders[i] != null) columnReaders[i].close();
+        // TODO: (ARROW NATIVE) handle tz, datetime & int96 rebase
```
https://github.com/apache/datafusion-comet/issues/1638

andygrove commented on common/src/main/java/org/apache/comet/parquet/NativeBatchReader.java:
```diff
@@ -263,111 +272,129 @@ public void init() throws URISyntaxException, IOException {
+          FileReader.filterRowGroups(readOptions, footer.getBlocks(), fileReader);
+      totalRowCount = fileReader.getFilteredRecordCount();
+      if (totalRowCount == 0) {
+        // all the data us filtered out.
```
```suggestion
        // all the data is filtered out.
```

andygrove reviewed (approved):
LGTM. Thanks @parthchandra!

parthchandra commented:
@andygrove I had introduced a regression which I fixed since your approval (if you want to look again)
Basically, the results of the call to `findRowIndexColumnIndexInSchema` were never being. used. Bit it had the side effect of throwing an exception in case of a incompatible type for row index columns being found in the schema and the unit test was checking for that specific exception. I've reintroduced that check.

andygrove commented on common/src/main/java/org/apache/comet/parquet/BatchReader.java:
```diff
@@ -282,6 +282,7 @@ public void init() throws URISyntaxException, IOException {
     // Initialize missing columns and use null vectors for them
     missingColumns = new boolean[columns.size()];
     List<String[]> paths = requestedSchema.getPaths();
+    ShimFileFormat.findRowIndexColumnIndexInSchema(sparkSchema);
```
would you mind adding a comment explaining that the result is unused but that the call has side effects?
perhaps adding `let _ = ` would help make this clear

andygrove replied:
oops, this is Java not Rust, so ignore the `let _ =` suggestion!

parthchandra replied:
I'll add a comment in the next PR.
